src/DelegateToken.sol

```diff
@@ -186,521 +162,254 @@ contract DelegateToken is ReentrancyGuard, Ownable2Step, ERC2981, IDelegateToken
-        IDelegateRegistry.Delegation[] memory delegation = IDelegateRegistry(delegateRegistry).getDelegationsFromHashes(delegationHash);
-        if (!(delegation[0].type_ == IDelegateRegistry.DelegationType.ERC721 && delegation[0].rights == "")) revert InvalidFlashloan();
-        IDelegateToken(delegation[0].contract_).transferFrom(address(this), receiver, delegation[0].tokenId);
-
-        if (INFTFlashBorrower(receiver).onFlashLoan{value: msg.value}(msg.sender, delegation[0].contract_, delegation[0].tokenId, data) != flashLoanCallBackSuccess) {
-            revert InvalidFlashloan();
+    function create(Structs.DelegateInfo calldata delegateInfo, uint256 salt) external nonReentrant returns (uint256 delegateTokenId) {
+        TransferHelpers.checkAndPullByType(erc1155PullAuthorization, delegateInfo);
+        Helpers.revertInvalidExpiry(delegateInfo.expiry);
+        Helpers.revertToIsZero(delegateInfo.delegateHolder);
+        delegateTokenId = Helpers.delegateId(msg.sender, salt);
+        StorageHelpers.revertAlreadyExisted(delegateTokenInfo, delegateTokenId);
+        StorageHelpers.incrementBalance(balances, delegateInfo.delegateHolder);
+        StorageHelpers.writeExpiry(delegateTokenInfo, delegateTokenId, delegateInfo.expiry);
+        emit Transfer(address(0), delegateInfo.delegateHolder, delegateTokenId);
+        bytes32 newRegistryHash = 0;
+        if (delegateInfo.tokenType == IDelegateRegistry.DelegationType.ERC721) {
+            newRegistryHash = RegistryHashes.erc721Hash(address(this), delegateInfo.rights, delegateInfo.delegateHolder, delegateInfo.tokenId, delegateInfo.tokenContract);
+            StorageHelpers.writeRegistryHash(delegateTokenInfo, delegateTokenId, newRegistryHash);
+            RegistryHelpers.delegateERC721(delegateRegistry, newRegistryHash, delegateInfo);
+        } else if (delegateInfo.tokenType == IDelegateRegistry.DelegationType.ERC20) {
+            StorageHelpers.writeUnderlyingAmount(delegateTokenInfo, delegateTokenId, delegateInfo.amount);
+            newRegistryHash = RegistryHashes.erc20Hash(address(this), delegateInfo.rights, delegateInfo.delegateHolder, delegateInfo.tokenContract);
+            StorageHelpers.writeRegistryHash(delegateTokenInfo, delegateTokenId, newRegistryHash);
+            RegistryHelpers.delegateERC20(delegateRegistry, newRegistryHash, delegateInfo);
+        } else if (delegateInfo.tokenType == IDelegateRegistry.DelegationType.ERC1155) {
+            StorageHelpers.writeUnderlyingAmount(delegateTokenInfo, delegateTokenId, delegateInfo.amount);
+            newRegistryHash =
+                RegistryHashes.erc1155Hash(address(this), delegateInfo.rights, delegateInfo.delegateHolder, delegateInfo.tokenId, delegateInfo.tokenContract);
+            StorageHelpers.writeRegistryHash(delegateTokenInfo, delegateTokenId, newRegistryHash);
+            RegistryHelpers.delegateERC1155(delegateRegistry, newRegistryHash, delegateInfo);
         }
```

Comment:
Let's throw an explicit error here if people try to create DelegateToken of type ALL or CONTRACT

Reply:
Looks like this is already handled in TransferHelpers.checkAndPullByType(). Resolving.